File: performance_validation/benchmarks/real_implementations/lexer_bench.cpp
// C++ Lexer Benchmark Implementation
#include <iostream>
#include <fstream>
#include <chrono>
#include <vector>
#include <string>
#include <cctype>
#include <unordered_set>

enum class TokenType {
    KEYWORD, IDENTIFIER, NUMBER, STRING, OPERATOR, PUNCTUATION, COMMENT, WHITESPACE, EOF_TOKEN
};

struct Token {
    TokenType type;
    std::string value;
    size_t line;
    size_t column;
};

class Lexer {
private:
    std::string source;
    size_t pos = 0;
    size_t line = 1;
    size_t column = 1;
    
    std::unordered_set<std::string> keywords = {
        "fun", "val", "var", "if", "else", "when", "for", "while",
        "class", "interface", "object", "return", "break", "continue",
        "true", "false", "null", "this", "super", "import", "package"
    };
    
    char current() {
        if (pos >= source.length()) return '\0';
        return source[pos];
    }
    
    char peek(int offset = 1) {
        size_t next_pos = pos + offset;
        if (next_pos >= source.length()) return '\0';
        return source[next_pos];
    }
    
    void advance() {
        if (pos < source.length()) {
            if (source[pos] == '\n') {
                line++;
                column = 1;
            } else {
                column++;
            }
            pos++;
        }
    }
    
    void skipWhitespace() {
        while (std::isspace(current())) {
            advance();
        }
    }
    
    Token scanIdentifier() {
        size_t start = pos;
        size_t start_col = column;
        
        while (std::isalnum(current()) || current() == '_') {
            advance();
        }
        
        std::string value = source.substr(start, pos - start);
        TokenType type = keywords.count(value) ? TokenType::KEYWORD : TokenType::IDENTIFIER;
        
        return {type, value, line, start_col};
    }
    
    Token scanNumber() {
        size_t start = pos;
        size_t start_col = column;
        
        while (std::isdigit(current())) {
            advance();
        }
        
        if (current() == '.' && std::isdigit(peek())) {
            advance(); // consume '.'
            while (std::isdigit(current())) {
                advance();
            }
        }
        
        return {TokenType::NUMBER, source.substr(start, pos - start), line, start_col};
    }
    
    Token scanString() {
        size_t start = pos;
        size_t start_col = column;
        char quote = current();
        advance(); // consume opening quote
        
        while (current() != quote && current() != '\0') {
            if (current() == '\\') {
                advance(); // skip escape char
                if (current() != '\0') advance(); // skip escaped char
            } else {
                advance();
            }
        }
        
        if (current() == quote) {
            advance(); // consume closing quote
        }
        
        return {TokenType::STRING, source.substr(start, pos - start), line, start_col};
    }
    
    Token scanComment() {
        size_t start = pos;
        size_t start_col = column;
        
        if (current() == '/' && peek() == '/') {
            // Single-line comment
            while (current() != '\n' && current() != '\0') {
                advance();
            }
        } else if (current() == '/' && peek() == '*') {
            // Multi-line comment
            advance(); advance(); // consume /*
            while (!(current() == '*' && peek() == '/') && current() != '\0') {
                advance();
            }
            if (current() == '*') {
                advance(); advance(); // consume */
            }
        }
        
        return {TokenType::COMMENT, source.substr(start, pos - start), line, start_col};
    }
    
public:
    Lexer(const std::string& src) : source(src) {}
    
    std::vector<Token> tokenize() {
        std::vector<Token> tokens;
        
        while (pos < source.length()) {
            skipWhitespace();
            
            if (pos >= source.length()) break;
            
            char ch = current();
            
            if (std::isalpha(ch) || ch == '_') {
                tokens.push_back(scanIdentifier());
            } else if (std::isdigit(ch)) {
                tokens.push_back(scanNumber());
            } else if (ch == '"' || ch == '\'') {
                tokens.push_back(scanString());
            } else if (ch == '/' && (peek() == '/' || peek() == '*')) {
                tokens.push_back(scanComment());
            } else {
                // Operators and punctuation
                size_t start_col = column;
                std::string op(1, ch);
                advance();
                
                // Check for multi-character operators
                if ((ch == '=' || ch == '!' || ch == '<' || ch == '>') && current() == '=') {
                    op += current();
                    advance();
                } else if ((ch == '&' && current() == '&') || (ch == '|' && current() == '|')) {
                    op += current();
                    advance();
                }
                
                tokens.push_back({TokenType::OPERATOR, op, line, start_col});
            }
        }
        
        tokens.push_back({TokenType::EOF_TOKEN, "", line, column});
        return tokens;
    }
};

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " <input_file> [iterations]" << std::endl;
        return 1;
    }
    
    std::string filename = argv[1];
    int iterations = argc > 2 ? std::atoi(argv[2]) : 1;
    
    // Read file
    std::ifstream file(filename);
    if (!file) {
        std::cerr << "Error: Cannot open file " << filename << std::endl;
        return 1;
    }
    
    std::string source((std::istreambuf_iterator<char>(file)),
                       std::istreambuf_iterator<char>());
    file.close();
    
    // Warmup
    for (int i = 0; i < 5; i++) {
        Lexer warmup_lexer(source);
        auto tokens = warmup_lexer.tokenize();
    }
    
    // Benchmark
    std::vector<double> times;
    size_t total_tokens = 0;
    
    for (int i = 0; i < iterations; i++) {
        auto start = std::chrono::high_resolution_clock::now();
        
        Lexer lexer(source);
        auto tokens = lexer.tokenize();
        
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        
        times.push_back(duration.count() / 1000000.0); // Convert to seconds
        total_tokens = tokens.size();
    }
    
    // Output results in JSON format
    std::cout << "{" << std::endl;
    std::cout << "  \"language\": \"cpp\"," << std::endl;
    std::cout << "  \"benchmark\": \"lexer\"," << std::endl;
    std::cout << "  \"iterations\": " << iterations << "," << std::endl;
    std::cout << "  \"tokens_processed\": " << total_tokens << "," << std::endl;
    std::cout << "  \"times\": [";
    
    for (size_t i = 0; i < times.size(); i++) {
        std::cout << times[i];
        if (i < times.size() - 1) std::cout << ", ";
    }
    
    std::cout << "]," << std::endl;
    
    // Calculate statistics
    double sum = 0;
    for (double t : times) sum += t;
    double avg = sum / times.size();
    
    double tokens_per_sec = total_tokens / avg;
    
    std::cout << "  \"average_time\": " << avg << "," << std::endl;
    std::cout << "  \"tokens_per_second\": " << tokens_per_sec << std::endl;
    std::cout << "}" << std::endl;
    
    return 0;
}

File: performance_validation/benchmarks/real_implementations/lexer_bench.rs
// Rust Lexer Benchmark Implementation
use std::fs;
use std::time::Instant;
use std::env;
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq)]
enum TokenType {
    Keyword,
    Identifier,
    Number,
    String,
    Operator,
    Punctuation,
    Comment,
    Whitespace,
    Eof,
}

#[derive(Debug, Clone)]
struct Token {
    token_type: TokenType,
    value: String,
    line: usize,
    column: usize,
}

struct Lexer {
    source: Vec<char>,
    pos: usize,
    line: usize,
    column: usize,
    keywords: HashSet<String>,
}

impl Lexer {
    fn new(source: String) -> Self {
        let mut keywords = HashSet::new();
        for kw in &[
            "fun", "val", "var", "if", "else", "when", "for", "while",
            "class", "interface", "object", "return", "break", "continue",
            "true", "false", "null", "this", "super", "import", "package",
        ] {
            keywords.insert(kw.to_string());
        }

        Lexer {
            source: source.chars().collect(),
            pos: 0,
            line: 1,
            column: 1,
            keywords,
        }
    }

    fn current(&self) -> Option<char> {
        if self.pos < self.source.len() {
            Some(self.source[self.pos])
        } else {
            None
        }
    }

    fn peek(&self, offset: usize) -> Option<char> {
        let next_pos = self.pos + offset;
        if next_pos < self.source.len() {
            Some(self.source[next_pos])
        } else {
            None
        }
    }

    fn advance(&mut self) {
        if self.pos < self.source.len() {
            if self.source[self.pos] == '\n' {
                self.line += 1;
                self.column = 1;
            } else {
                self.column += 1;
            }
            self.pos += 1;
        }
    }

    fn skip_whitespace(&mut self) {
        while let Some(ch) = self.current() {
            if ch.is_whitespace() {
                self.advance();
            } else {
                break;
            }
        }
    }

    fn scan_identifier(&mut self) -> Token {
        let start = self.pos;
        let start_col = self.column;

        while let Some(ch) = self.current() {
            if ch.is_alphanumeric() || ch == '_' {
                self.advance();
            } else {
                break;
            }
        }

        let value: String = self.source[start..self.pos].iter().collect();
        let token_type = if self.keywords.contains(&value) {
            TokenType::Keyword
        } else {
            TokenType::Identifier
        };

        Token {
            token_type,
            value,
            line: self.line,
            column: start_col,
        }
    }

    fn scan_number(&mut self) -> Token {
        let start = self.pos;
        let start_col = self.column;

        while let Some(ch) = self.current() {
            if ch.is_numeric() {
                self.advance();
            } else {
                break;
            }
        }

        if self.current() == Some('.') && self.peek(1).map_or(false, |c| c.is_numeric()) {
            self.advance(); // consume '.'
            while let Some(ch) = self.current() {
                if ch.is_numeric() {
                    self.advance();
                } else {
                    break;
                }
            }
        }

        let value: String = self.source[start..self.pos].iter().collect();
        Token {
            token_type: TokenType::Number,
            value,
            line: self.line,
            column: start_col,
        }
    }

    fn scan_string(&mut self) -> Token {
        let start = self.pos;
        let start_col = self.column;
        let quote = self.current().unwrap();
        self.advance(); // consume opening quote

        while let Some(ch) = self.current() {
            if ch == quote {
                self.advance(); // consume closing quote
                break;
            } else if ch == '\\' {
                self.advance(); // skip escape char
                if self.current().is_some() {
                    self.advance(); // skip escaped char
                }
            } else {
                self.advance();
            }
        }

        let value: String = self.source[start..self.pos].iter().collect();
        Token {
            token_type: TokenType::String,
            value,
            line: self.line,
            column: start_col,
        }
    }

    fn scan_comment(&mut self) -> Token {
        let start = self.pos;
        let start_col = self.column;

        if self.current() == Some('/') && self.peek(1) == Some('/') {
            // Single-line comment
            while let Some(ch) = self.current() {
                if ch == '\n' {
                    break;
                }
                self.advance();
            }
        } else if self.current() == Some('/') && self.peek(1) == Some('*') {
            // Multi-line comment
            self.advance(); // consume /
            self.advance(); // consume *
            while self.current().is_some() {
                if self.current() == Some('*') && self.peek(1) == Some('/') {
                    self.advance(); // consume *
                    self.advance(); // consume /
                    break;
                }
                self.advance();
            }
        }

        let value: String = self.source[start..self.pos].iter().collect();
        Token {
            token_type: TokenType::Comment,
            value,
            line: self.line,
            column: start_col,
        }
    }

    fn tokenize(&mut self) -> Vec<Token> {
        let mut tokens = Vec::new();

        while self.pos < self.source.len() {
            self.skip_whitespace();

            if self.pos >= self.source.len() {
                break;
            }

            if let Some(ch) = self.current() {
                if ch.is_alphabetic() || ch == '_' {
                    tokens.push(self.scan_identifier());
                } else if ch.is_numeric() {
                    tokens.push(self.scan_number());
                } else if ch == '"' || ch == '\'' {
                    tokens.push(self.scan_string());
                } else if ch == '/' && (self.peek(1) == Some('/') || self.peek(1) == Some('*')) {
                    tokens.push(self.scan_comment());
                } else {
                    // Operators and punctuation
                    let start_col = self.column;
                    let mut op = String::from(ch);
                    self.advance();

                    // Check for multi-character operators
                    if (ch == '=' || ch == '!' || ch == '<' || ch == '>') 
                        && self.current() == Some('=') {
                        op.push(self.current().unwrap());
                        self.advance();
                    } else if (ch == '&' && self.current() == Some('&')) 
                           || (ch == '|' && self.current() == Some('|')) {
                        op.push(self.current().unwrap());
                        self.advance();
                    }

                    tokens.push(Token {
                        token_type: TokenType::Operator,
                        value: op,
                        line: self.line,
                        column: start_col,
                    });
                }
            }
        }

        tokens.push(Token {
            token_type: TokenType::Eof,
            value: String::new(),
            line: self.line,
            column: self.column,
        });

        tokens
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    
    if args.len() < 2 {
        eprintln!("Usage: {} <input_file> [iterations]", args[0]);
        std::process::exit(1);
    }
    
    let filename = &args[1];
    let iterations = if args.len() > 2 {
        args[2].parse().unwrap_or(1)
    } else {
        1
    };
    
    // Read file
    let source = fs::read_to_string(filename)
        .expect(&format!("Error: Cannot read file {}", filename));
    
    // Warmup
    for _ in 0..5 {
        let mut warmup_lexer = Lexer::new(source.clone());
        let _ = warmup_lexer.tokenize();
    }
    
    // Benchmark
    let mut times = Vec::new();
    let mut total_tokens = 0;
    
    for _ in 0..iterations {
        let start = Instant::now();
        
        let mut lexer = Lexer::new(source.clone());
        let tokens = lexer.tokenize();
        
        let duration = start.elapsed();
        
        times.push(duration.as_secs_f64());
        total_tokens = tokens.len();
    }
    
    // Output results in JSON format
    println!("{{");
    println!("  \"language\": \"rust\",");
    println!("  \"benchmark\": \"lexer\",");
    println!("  \"iterations\": {},", iterations);
    println!("  \"tokens_processed\": {},", total_tokens);
    print!("  \"times\": [");
    
    for (i, time) in times.iter().enumerate() {
        print!("{}", time);
        if i < times.len() - 1 {
            print!(", ");
        }
    }
    
    println!("],");
    
    // Calculate statistics
    let sum: f64 = times.iter().sum();
    let avg = sum / times.len() as f64;
    
    let tokens_per_sec = total_tokens as f64 / avg;
    
    println!("  \"average_time\": {},", avg);
    println!("  \"tokens_per_second\": {}", tokens_per_sec);
    println!("}}");
}